File: app/javascript/channels/presence_channel.js
import consumer from "./consumer"

const presenceChannel = consumer.subscriptions.create("PresenceChannel", {
  connected() {
    console.log("PresenceChannel에 연결되었습니다.");
  },

  disconnected() {
    console.log("PresenceChannel 연결이 해제되었습니다.");
  },

  received(data) {
    if (data.type === "full_update") {
      // 전체 온라인 사용자 목록을 새로 구성
      const onlineUsersEl = document.getElementById("online-users");
      onlineUsersEl.innerHTML = ""; // 기존 목록 초기화
      data.users.forEach(user => {
        if (!document.getElementById("user-" + user)) {
          const li = document.createElement("li");
          li.textContent = user;
          li.id = "user-" + user;
          onlineUsersEl.appendChild(li);
        }
      });
    } else if (data.type === "update") {
      // 개별 업데이트 처리 (앞서 구현한 코드와 동일)
      const onlineUsersEl = document.getElementById("online-users");
      if (data.status === "online") {
        if (!document.getElementById("user-" + data.user)) {
          const li = document.createElement("li");
          li.textContent = data.user;
          li.id = "user-" + data.user;
          onlineUsersEl.appendChild(li);
        }
      } else if (data.status === "offline") {
        const userEl = document.getElementById("user-" + data.user);
        if (userEl) {
          userEl.remove();
        }
      }
    }
  }  
});

export default presenceChannel;

File: app/channels/presence_channel.rb
class PresenceChannel < ApplicationCable::Channel
  def subscribed
    stream_from "presence_channel"
    OnlineUserStore.add(current_user.email)
    # 모든 클라이언트에 전체 목록 갱신 메시지 전송
    ActionCable.server.broadcast("presence_channel", { users: OnlineUserStore.all, type: "full_update" })
    # 자신의 상태 업데이트
    ActionCable.server.broadcast("presence_channel", { user: current_user.email, status: "online", type: "update" })
  end

  def unsubscribed
    # 연결 해제 시 온라인 사용자 목록에서 제거
    OnlineUserStore.remove(current_user.email)
    ActionCable.server.broadcast("presence_channel", { user: current_user.email, status: "offline", type: "update" })
  end
end
